Add a paragraph on integration tests (#203)
llvm/mlir-www: website/content/getting_started/TestingGuide.md

cmake --build . --target check-mlir-integration
```

Note that integration tests are relatively expensive to run (primarily due to
JIT compilation), and tend to be trickier to debug (with multiple compilation
steps _integrated_, it usually takes a bit of triaging to find the root cause
of a failure). We reserve e2e tests for cases that are hard to verify
otherwise, e.g. when composing and testing complex compilation pipelines. In
those cases, verifying run-time output tends to be easier then the checking
e.g. LLVM IR with FileCheck. Lowering optimized `linalg.matmul` (with tiling
and vectorization) is a good example. For less involved lowering pipelines or

> when there's almost 1-1 mapping between an Op and it's LLVM IR counterpart
> (e.g. `arith.cmpi` and LLVM IR `icmp` instruction), use regular tests (not
> requiring JIT compilation) instead.

Reviewer: ```suggestion
when there's almost 1-1 mapping between an Op and it's LLVM IR counterpart
(e.g. `arith.cmpi` and LLVM IR `icmp` instruction),  regular unit tests are considered
enough.
```

cmake --build . --target check-mlir-integration
```

Note that integration tests are relatively expensive to run (primarily due JIT
compilation), and hence we reserve e2e tests for verifying complex compilation
pipelines. Use them when e.g. lowering high-level operations like
`linalg.matmul` through involved transformations/optimizations such as tiling,
vectorization, progressive vector lowering or other. For less involved lowering
pipelines or when there's almost 1-1 mapping between an Op and it's LLVM IR
counterpart (e.g. `arith.cmpi` and LLVM IR `icmp` instruction), use regular
tests (not requiring JIT compilation) instead.

> 

Reviewer: The cost of JIT / execution is only part of the problem: the coupling of the entire stack can create low-signal / high-noise tests. Also when a test fails, it's harder to debug (is the problem with the test itself? Was it working "by luck"? Is the issue in any of the other layering and the change is just "unlucky" to expose a latent bug?) ; so the maintenance cost is harder than unit-tests.

Reviewer: > These are good points, but it's also getting a bit into "integration vs unit tests" 

Right, but isn't it what this entire new paragraph is about anyway?
